### PR TITLE
Update ClientPipelineApi, HttpMessageApi abstraction

### PR DESCRIPTION
**Pull request by live1206**
- Remove unused `ClientPipelineApi.CreateMessage`
- Add `HttpRequestOptionsApi` to `ClientPipelineApi.Send(Async)`
- Replace `HttpMessageApi.ResponseClassifier` with `HttpMessageApi.ApplyResponseClassifier`
- Make abstraction more generic

This fix is needed for below issues of Azure plugin:
https://github.com/Azure/azure-sdk-for-net/issues/46899
https://github.com/Azure/azure-sdk-for-net/issues/46900
https://github.com/Azure/azure-sdk-for-net/issues/46901
https://github.com/Azure/azure-sdk-for-net/issues/46902

### CODE REVIEW

**Comment by live1206** [comment]
Corresponding changes in Azure Plugin: https://github.com/Azure/azure-sdk-for-net/pull/46835

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/test/Providers/Abstractions/ClientPipelineApiTests.cs [hunk 1]
```diff
@@ -28,7 +29,7 @@ public void ValidateBodyOfRestClientOperationIsOverridden()
 
             Assert.NotNull(method);
             Assert.NotNull(method!.BodyStatements);
-            var test = method?.BodyStatements?.ToDisplayString();
+            var test = method!.BodyStatements!.ToDisplayString();
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/test/Providers/Abstractions/ClientPipelineApiTests.cs:32]: is this used? If not can we please remove it?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/ClientPipelineApi.cs [hunk 2]
```diff
@@ -18,13 +19,11 @@ protected ClientPipelineApi(Type type, ValueExpression original) : base(type, or
         {
         }
 
-        public abstract HttpMessageApi CreateMessage();
+        public abstract ValueExpression CreateMessage(ParameterProvider requestOptions, ValueExpression responseClassifier);
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/ClientPipelineApi.cs:22]: Why was this changed from a type of HttpRequestOptionsApi to ParameterProvider?

Comment (live1206) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/ClientPipelineApi.cs:22]: you are right, it's better to have it as HttpRequestOptionsApi

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/test/Providers/Abstractions/ClientPipelineApiTests.cs (hunk 1 above)
Comment (live1206) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/test/Providers/Abstractions/ClientPipelineApiTests.cs]: remvoed

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/ClientPipelineApi.cs (hunk 2 above)
Comment (live1206) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/ClientPipelineApi.cs]: updated

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/HttpMessageApi.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ protected HttpMessageApi(Type type, ValueExpression original) : base(type, origi
 
         public abstract ValueExpression BufferResponse();
 
-        public abstract ValueExpression ResponseClassifier();
+        public abstract MethodBodyStatement Assign(StatusCodeClassifierApi statusCodeClassifier);
```
Comment (live1206) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/HttpMessageApi.cs:24]: This is due to this assignment is not needed for Azure.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/HttpMessageApi.cs [hunk 4]
```diff
@@ -21,7 +21,7 @@ protected HttpMessageApi(Type type, ValueExpression original) : base(type, origi
...
 
         public abstract MethodBodyStatement Apply(ValueExpression options);
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/HttpMessageApi.cs:26]: What is this meant to Apply? Is this the RequestContext? Can the parameter be of type HttpRequestOptionsApi?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/HttpMessageApi.cs (hunk 3 above)
Comment (live1206) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/HttpMessageApi.cs:24]: sure, it can be AssignResponseClassifier

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/HttpMessageApi.cs (hunk 4 above)
Comment (live1206) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/HttpMessageApi.cs:26]: No, it's not RequestContext as we rename this to AssignResponseClassifier, the value applying here is StatusCodeClassifierApi

Comment (live1206) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/Abstractions/HttpMessageApi.cs]: renamed to SetRequestContext and the input type should be HttpRequestOptionsApi

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/RestClientProvider.cs [hunk 5]
```diff
@@ -165,18 +165,18 @@ private MethodProvider BuildCreateRequestMethod(InputOperation operation)
                 signature,
                 new MethodBodyStatements(
                 [
-                    Declare("message", pipelineField.CreateMessage().ToApi<HttpMessageApi>(), out HttpMessageApi message),
-                    message.ResponseClassifier().Assign(classifier).Terminate(),
-                    Declare("request", message.Request().ToApi<HttpRequestApi>(), out HttpRequestApi request), // ScopedApi<PipelineRequest>
-                    request.SetMethod(operation.HttpMethod).Terminate(),
+                    Declare("message", pipelineField.CreateMessage(options.AsExpression.ToApi<HttpRequestOptionsApi>(), classifier).ToApi<HttpMessageApi>(), out HttpMessageApi message),
+                    message.ApplyResponseClassifier(((MemberExpression)classifier).ToApi<StatusCodeClassifierApi>()),
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/RestClientProvider.cs:169]: Why do we need the cast to MemberExpression?

Comment (live1206) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/RestClientProvider.cs:169]: Because we need to cast from PropertyProvider to StatusCodeClassifierApi, and we need to cast PropertyProvider to ValueExpression to be able to call ToApi<HttpRequestApi>()

Comment (live1206) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/RestClientProvider.cs]: Removed the casting with adding overload methods of ToApi

**Review by JoshLove-msft — APPROVED** [review]
